cmd/goal/box.go

```diff
@@ -0,0 +1,112 @@
+// Copyright (C) 2019-2022 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package main
+
+import (
+	"encoding/base64"
+
+	"github.com/spf13/cobra"
+)
+
+var boxName string
+
+func init() {
+	appCmd.AddCommand(appBoxCmd)
+
+	appBoxCmd.AddCommand(appBoxInfoCmd)
+	appBoxCmd.AddCommand(appBoxListCmd)
+	appBoxCmd.PersistentFlags().Uint64Var(&appIdx, "app-id", 0, "Application ID")
+	appBoxCmd.MarkFlagRequired("app-id")
+
+	appBoxInfoCmd.Flags().StringVarP(&boxName, "name", "n", "", "Application box name. Use the same form as app-arg to name the box.")
+	appBoxInfoCmd.MarkFlagRequired("name")
+}
+
+var appBoxCmd = &cobra.Command{
+	Use:   "box",
+	Short: "Read application box data",
+	Args:  cobra.ArbitraryArgs,
+	Run: func(cmd *cobra.Command, args []string) {
+		// If no arguments passed, we should fallback to help
+		cmd.HelpFunc()(cmd, args)
+	},
+}
+
+var appBoxInfoCmd = &cobra.Command{
+	Use:   "info",
+	Short: "Retrieve information about an application box.",
+	Long:  "Retrieve information about an application box. The returned box name and value are base64 encoded.",
+	Args:  validateNoPosArgsFn,
+	Run: func(cmd *cobra.Command, args []string) {
+		_, client := getDataDirAndClient()
+
+		// Ensure box name is specified
+		if boxName == "" {
+			reportErrorf(errorMissingBoxName)
+		}
+
+		// Get box info
+		box, err := client.GetApplicationBoxByName(appIdx, boxName)
+		if err != nil {
+			reportErrorf(errorRequestFail, err)
+		}
+
+		// Print box info
+		encodedName := base64.StdEncoding.EncodeToString(box.Name)
+		encodedValue := base64.StdEncoding.EncodeToString(box.Value)
+		reportInfof("Name:  %s", encodedName)
+		reportInfof("Value: %s", encodedValue)
```

Comment:
Why not using your fancy `encodeValueInAppCallBytesFormat` here?

Reply:
I'm not opposed to doing that. I thought it made sense to base64 encode here for consistency's sake. In other words, the values you get are always base64 encoded and you wouldn't require an additional step to consider what byte format you received. The counterpoint is that you likely aren't using this command to test box values in a programmatic way, so it would be convenient if it displayed unicode strings conveniently.

My assumption was that the app call bytes format was most valuable for box names as those are likely human-readable. Since you must pass the box name into this command, there wouldn't be much added value in repeating it back. It also avoids the slightly confusing situation of passing a name such as `b64:bmFtZQ==` and seeing the name `str:name` returned.

Reply:
Yeah, I can see that.  I'm ok either way.  I'll tag a couple of people for opinions, but I think this is fine. @michaeldiamant  @jasonpaulos @winder 

I'll merge if no objections.

Here are some alternatives to consider.

1. For name, use the encoding that was given on the command-line (after checking and screaming if they don't match!)
2. For value, keep as is, but prefix with "b64:" as a signal.

Reply:
Failing a test on circle.

Reply:
Yep. Still working on these changes.

Reply:
Since my feedback was requested:  I don't feel strongly though I suggest matching `list` output by using `encodeValueInAppCallBytesFormat`.

From my perspective, since we're providing a general purpose retrieval method, criteria to consider include:
* Provide a consistent response structure for related commands.
* Take into account prevailing `goal` command response conventions.  I didn't spot any related concerns to consider.

Reply:
Okay. I've changed the app box info command to return name and value in app call bytes format. I included @jannotti's suggestion of using whatever box name is passed in.